### PR TITLE
E2E test workflow for GCP enclave

### PR DESCRIPTION
**Pull request by yishi-ttd**
Check e2e/README.md for details.

### CODE REVIEW

File: e2e/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+You could also leverage them to bring up a local docker-compose cluster contains:
+ - localstack (local S3)
+ - core (depends on localstack)
+ - optout (depends on localstack and core)
+
+and expose public Urls via ngrok, which could be used for private operator test.
+
+# How to run locally
+Set below config in `./e2e/.e2e.sh`
```
Comment (lunwang-ttd) [inline e2e/README.md:13]:
```suggestion
Set below config in `./e2e/e2e.sh`
```

File: e2e/docker/uid2-optout/conf/local-e2e-docker-config.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  "optout_replica_uris": null,
+  "partners_metadata_path": "<NGROK_URL_CORE>/partners/refresh",
+  "operators_metadata_path": "<NGROK_URL_CORE>/operators/refresh",
+  "core_attest_url": "<NGROK_URL_CORE>/attest",
+  "core_api_token": "UID2-O-L-127-pDqphU.6FuXzThQMY8YEsCA8crqvAlzyGrjcF8P6XO84=",
+  "aws_kms_jwt_signing_key_id": "ff275b92-0def-4dfc-b0f6-87c96b26c6c7",
+  "aws_kms_jwt_signing_public_keys": "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA9tXqXNeeuA3lmfNF2AF06OLY6icxEOlXaZEXyh62ieFjakvDeR438s5gA1ToKqpW5uTPkKtP1YjHf16ofjlDYcLxCKOombjW1m2t/3FFfWRaZgr/SFr5TaI3N86tAvf313kql8GVrre/HQLhaHUXQJxDYh3/Wl7+KSMrMwMQCwkgnf1AmGUcPl1AzxuKUia8IM9joZKq9UpxRdWkLeqJ47d0bX5+SFLOwRhFxnejNx3iJYpEi43aFxLnr1WmTtMrIkHYmA2GENnhXsrV3n0bqLUah+HS6S+m3lBUW09TEFbagtktCtsFqvE+YExXo8b+hOnlakQtsbtJY0xg8Z1AtwIDAQAB,MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA9tXqXNeeuA3lmfNF2AF06OLY6icxEOlXaZEXyh62ieFjakvDeR438s5gA1ToKqpW5uTPkKtP1YjHf16ofjlDYcLxCKOombjW1m2t/3FFfWRaZgr/SFr5TaI3N86tAvf313kql8GVrre/HQLhaHUXQJxDYh3/Wl7+KSMrMwMQCwkgnf1AmGUcPl1AzxuKUia8IM9joZKq9UpxRdWkLeqJ47d0bX5+SFLOwRhFxnejNx3iJYpEi43aFxLnr1WmTtMrIkHYmA2GENnhXsrV3n0bqLUah+HS6S+m3lBUW09TEFbagtktCtsFqvE+YExXo8b+hOnlakQtsbtJY0xg8Z1AtwIDAQAB",
```
Comment (lunwang-ttd) [inline e2e/docker/uid2-optout/conf/local-e2e-docker-config.json:20]: Is this a real key used by AWS KMS? Do we need it for e2e testing?

File: e2e/start_docker.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  echo "NGROK_URL_OPTOUT can not be empty"
+  exit 1
+fi
+
+# replace placeholders
+sed -i.bak "s#<CORE_VERSION>#$CORE_VERSION#g" $COMPOSE_FILE
+sed -i.bak "s#<OPTOUT_VERSION>#$OPTOUT_VERSION#g" $COMPOSE_FILE
+
+#cat <<< $(jq '(.aws_s3_endpoint) |='\"$NGROK_URL_LOCALSTACK\"'' $CORE_CONFIG_FILE) > $CORE_CONFIG_FILE
+#cat <<< $(jq '(.aws_s3_endpoint) |='\"$NGROK_URL_LOCALSTACK\"'' $OPTOUT_CONFIG_FILE) > $OPTOUT_CONFIG_FILE
```
Comment (lunwang-ttd) [inline e2e/start_docker.sh:41]: Remove unused codes?

**Review by lunwang-ttd — APPROVED** [review]
LGTM

**Comment by vishalegbert-ttd** [comment]
Can we checkout Core and Opt Out repos and use the confs from the repos rather than duplicating it here?

Ideally we should do the same for Localstack fake data as well, but we'll probably need to move dev workspace repo to GitHub first

File: e2e/docker/uid2-optout/conf/local-e2e-docker-config.json (hunk 2 above)
Comment (yishi-ttd) [reply e2e/docker/uid2-optout/conf/local-e2e-docker-config.json:20]: I don't think so. I copied it from OPTOUT repo local config.

**Comment by vishalegbert-ttd** [comment]
I think we can move localstack fake data into a public repo in UnifiedID2 and check that out instead. dev-workspace can reference the same data as a submodule

File: scripts/gcp-oidc/entrypoint.sh [hunk 4]
```diff
@@ -34,6 +34,15 @@ if [ $? -ne 0 ]; then
   exit 1
 fi
 
+# -- replace base URLs if both CORE_BASE_URL and OPTOUT_BASE_URL are provided
+# -- using hardcoded domains is fine because they should not be changed frequently
+if [ -n "${CORE_BASE_URL}" -a -n "${OPTOUT_BASE_URL}" -a "${DEPLOYMENT_ENVIRONMENT}" != 'prod' ]; then
```
Comment (vishalegbert-ttd) [inline scripts/gcp-oidc/entrypoint.sh:39]: Does this only work in debug mode?

**Comment by yishi-ttd** [comment]
> Can we checkout Core and Opt Out repos and use the confs from the repos rather than duplicating it here?
> 
> Ideally we should do the same for Localstack fake data as well, but we'll probably need to move dev workspace repo to GitHub first

Sounds good.
I'd like to merge this PR first, then address it:  I need to check in configs (with NGROK placeholder) to core/optout repos first then update this flow.

File: scripts/gcp-oidc/entrypoint.sh (hunk 4 above)
Comment (yishi-ttd) [reply scripts/gcp-oidc/entrypoint.sh:39]: We only allow it to work in integration env since we don't want to allow endpoint override in prod.

**Comment by yishi-ttd** [comment]
> I think we can move localstack fake data into a public repo in UnifiedID2 and check that out instead. dev-workspace can reference the same data as a submodule

Yes, that would be ideal and could reduce the maintain overhead.

**Comment by vishalegbert-ttd** [comment]
>I'd like to merge this PR first, then address it: I need to check in configs (with NGROK placeholder) to core/optout repos first then update this flow.

Sure

Could we use jq to update it in the pipeline, hopefully we don't have to commit the ngrok placeholder into core and optout?

**Comment by yishi-ttd** [comment]
Will merge this first then make changes based on it.

**Pull request merged**